src/transformers/generation_utils.py

```diff
@@ -799,7 +929,19 @@ def greedy_search(
             # increase cur_len
             cur_len = cur_len + 1
 
-        return input_ids
```

Comment:
Ok, now we have to add the ugly hack to keep backwards compatibility! Would be nice to get @sgugger and @LysandreJik input here as well. 

Context:
Similar to the model's forward method we should add `GenerationOutputs` to generate. It's way more problematic to handle the backwards compatibility here though IMO. As one can see in the above line, we previously just returned a single tensor `input_ids`. Now with the `GenerationOutputs` we have to change this to a "tuple-like" output. The question now is how to implement this feature without **any** breaking backward compatibility since `generate()` is such a public method. It's not sufficient to just check return `input_ids` when all `output_attentions`, `output_hidden_states`, `output_scores` are false and return the `GenerationOutputs` otherwise because people might previously had `config.output_attentions` set to `True`. In the long run, I want to force `generate()` to always return a `GenerationOutput`, but for now I think we have to add some flag like `return_generation_dict` only to the `generate` function that defaults to `False` and that has to be set to `True` to return the new `GenerationOutput` class. Then we should update all our examples to have the flag set to `True` and show the new API. In a next step we could add a depreciation warning that somewhat forcing the user to set `return_generation_dict` to `True` and eventually we delete this flag. Not sure if that's the best approach....keen to hear your thoughts @sgugger, @LysandreJik, @SBrandeis

Reply:
I agree with @patrickvonplaten. For additional context @SBrandeis, we added the model outputs in stages like Patrick mentioned with a `return_dict` flag. At first it defaulted to `False` but we manually set it to `True` in all our examples/tests. Then we switched the default during a major release (to :obj:`True`). And maybe in a future major release we will remove it entirely.

I think we need to take the same approach here for backward compatibility, so I agree with your proposal @patrickvonplaten (except we'll deprecate and not depreciate :-p )

Reply:
Cool! So the best idea is to probably also just add a `config.return_generation_output` flag no?

Haha, yeah I really need to deprecate the word depreciate in my vocab

Reply:
Not sure `return_generation_output` is the right name, but that's the idea yes.

Reply:
Okey then let's go for a `config.return_generation_output` (we can also think about a better name before merging).

We can add a `if return_generation_outut: ` then here @SBrandeis

Reply:
It's more of a `return_dict_in_generate` than `return_generation_output`. I do expect the generation to return an output whatever my args are ;-)

Reply:
Also agree with both of you, backwards compatibility is incredibly important here given how public facing this is. Having the same approach as we have had with the `return_dict` works for me!